tests/transforms/test_control.py

```diff
@@ -0,0 +1,192 @@
+from functools import partial
+import numpy as np
+import pennylane as qml
+from pennylane.tape import QuantumTape
+from pennylane.transforms.control import ctrl, ControlledOperation
+from pennylane.tape.tape import expand_tape
+
+
+def assert_equal_operations(ops1, ops2):
+    assert len(ops1) == len(ops2)
+    for op1, op2 in zip(ops1, ops2):
+        assert type(op1) == type(op2)
+        assert op1.wires == op2.wires
+        np.testing.assert_allclose(op1.parameters, op2.parameters)
+
+
+def test_control_sanity_check():
+
+    def make_ops():
+        qml.RX(0.123, wires=0)
+        qml.RY(0.456, wires=2)
+        qml.RX(0.789, wires=0)
+        qml.PauliX(wires=2)
+        qml.PauliY(wires=4)
+        qml.PauliZ(wires=0)
+
+    with QuantumTape() as tape:
+        cmake_ops = ctrl(make_ops, control=1)
+        #Execute controlled version.
+        cmake_ops()
+
+    expected = [
+        qml.CRX(0.123, wires=[1, 0]),
+        qml.CRY(0.456, wires=[1, 2]),
+        qml.CRX(0.789, wires=[1, 0]),
+        qml.CNOT(wires=[1, 2]),
+        qml.CY(wires=[1, 4]),
+        qml.CZ(wires=[1, 0]),
+    ]
+    assert len(tape.operations) == 1
+    ctrl_op = tape.operations[0]
+    assert isinstance(ctrl_op, ControlledOperation)
+    assert len(ctrl_op.subtape.operations) == 6
+    expanded = ctrl_op.expand()
+    assert_equal_operations(expanded.operations, expected)
+
+
+def test_adjoint_of_control():
+
+    def my_op(a, b, c):
+        qml.RX(a, wires=2)
+        qml.RY(b, wires=3)
+        qml.RZ(c, wires=0)
+
+    with QuantumTape() as tape1:
+        cmy_op_dagger = qml.adjoint(ctrl(my_op, 5))
+        # Execute controlled and adjointed version of my_op.
+        cmy_op_dagger(0.789, 0.123, c=0.456)
+
+    with QuantumTape() as tape2:
+        cmy_op_dagger = ctrl(qml.adjoint(my_op), 5)
+        # Execute adjointed and controlled version of my_op.
+        cmy_op_dagger(0.789, 0.123, c=0.456)
+
+    expected = [
+        qml.CRZ(-0.456, wires=[5, 0]),
+        qml.CRY(-0.123, wires=[5, 3]),
+        qml.CRX(-0.789, wires=[5, 2]),
+    ]
+    for tape in [tape1, tape2]:
+        assert len(tape.operations) == 1
+        ctrl_op = tape.operations[0]
+        assert isinstance(ctrl_op, ControlledOperation)
+        expanded = ctrl_op.expand()
+        assert_equal_operations(expanded.operations, expected)
+
+def test_nested_control():
+    with QuantumTape() as tape:
+        CCX = ctrl(ctrl(qml.PauliX, 7), 3)
+        CCX(wires=0)
+    assert len(tape.operations) == 1
+    op = tape.operations[0]
+    assert isinstance(op, ControlledOperation)

```

Comment:
So if you ask for the control of a non-registered op, you get a (potentially nested) `ControlledOperation`?

Reply:
No, if you ask for a control of a non-registered op, then nothing will happen unless the operation is expanded. In that case, a `NotimplementedError` is raised.

Reply:
What if you have a non-registered op, but when expanded, all resulting ops _are_ registered?

Reply:
Good point. Let me see if I can add that...

Reply:
Any updates on this thread?

Reply:
Added. It's kinda awesome ngl, good call.